silence/silence_test.go

```diff
@@ -203,27 +207,25 @@ func TestSilences_Maintenance_DefaultMaintenanceFuncDoesntCrash(t *testing.T) {
 func TestSilences_Maintenance_SupportsCustomCallback(t *testing.T) {
 	f, err := ioutil.TempFile("", "snapshot")
 	require.NoError(t, err, "creating temp file failed")
-	s := &Silences{st: state{}, logger: log.NewNopLogger(), now: utcNow, metrics: newMetrics(nil, nil)}
+	clock := clock.NewMock()
+	s := &Silences{st: state{}, logger: log.NewNopLogger(), clock: clock, metrics: newMetrics(nil, nil)}
 	stopc := make(chan struct{})
-	var mtx sync.Mutex
-	var mc int
 
+	calls := atomic.NewInt32(0)
 	go s.Maintenance(100*time.Millisecond, f.Name(), stopc, func() (int64, error) {
-		mtx.Lock()
-		mc++
-		mtx.Unlock()
-
+		calls.Add(1)
 		return 0, nil
 	})
+	runtime.Gosched()
+
+	clock.Add(100 * time.Millisecond)
 
-	time.Sleep(200 * time.Millisecond)
+	// Stop the maintenance loop. We should get exactly one more execution of the maintenance func.
 	close(stopc)
 
 	require.Eventually(t, func() bool {
-		mtx.Lock()
```

Comment:
I was surprised to see the use of an `eventually` here now that we have a mock-able clock. You can get rid of the eventually call by making sure you wait until the maintenance function returns:

```golang
func TestSilences_Maintenance_SupportsCustomCallback(t *testing.T) {
	f, err := ioutil.TempFile("", "snapshot")
	require.NoError(t, err, "creating temp file failed")
	clock := clock.NewMock()
	s := &Silences{st: state{}, logger: log.NewNopLogger(), clock: clock, metrics: newMetrics(nil, nil)}
	stopc := make(chan struct{})

	calls := atomic.NewInt32(0)
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		s.Maintenance(100*time.Millisecond, f.Name(), stopc, func() (int64, error) {
			calls.Add(1)
			return 0, nil
		})
		wg.Done()
	}()
	runtime.Gosched()

	clock.Add(100 * time.Millisecond)

	// Stop the maintenance loop. We should get exactly one more execution of the maintenance func.
	close(stopc)
	<-stopc
	wg.Wait()

	require.EqualValues(t, calls.Load(), 2)
}
```

Reply:
Definitely, a nit and can be addressed as part of a second PR if you wish.